### DISCUSSION TITLE
FastAPI integration with OIDC  and SSO

### DISCUSSION THREAD

**Discussion by karenheg** [fastapi/fastapi#9131] [Questions]
Hello I'm trying to implement a way to Verify and decrypt OIDC tokens.

I've reviewed [Authlib](https://docs.authlib.org/en/stable/client/fastapi.html) documentation and seems google authentication is pretty easy but what if I have a different provider? 

I saw this pice of code 
`@app.get("/login")
def login_via_google(request: Request):
    redirect_uri = 'https://example.com/auth'
    return await oauth.google.authorize_redirect(request, redirect_uri)` 

and also for twitter but if in my case is a SSO provider how can I configured it or how will be the `oauth.somehting.authorize_redirect`?

Or do you know a better approach to implement OIDC with FastAPI? I'm using a SSO provider

Thanks Guys

**Comment by gucharbon** [comment]
If you have your own SSO which supports OIDC and expose a well-known URL, then you can juste copy/paste this example and update the variable [`CONF_URL`](https://github.com/authlib/demo-oauth-client/blob/master/fastapi-google-login/app.py#L15). (Authlib documentation does show how to use google provider, but the same things could be accomplished with your own provider).

https://github.com/authlib/demo-oauth-client/blob/master/fastapi-google-login/app.py

You can see that creating your own oauth provider is straightforward. For example using a local Keycloak instance you would do:

```python
CONF_URL = 'http://localhost:8080/auth/realms/YOUR_REALM/.well-known/openid-configuration'
oauth.register(
    name='keycloak_sso',
    server_metadata_url=CONF_URL,
    client_kwargs={
        'scope': 'openid email profile'
    }
)
```
You can then access this provider using `oauth.keycloak_sso`, and perform any authentication as mentioned in the documentation.

If you'd like to issue token yourself, you can easily obtain them from your SSO now:

```python
@app.post("/token")
def authenticate_user_password(username: str, password: str):
    """Authenticate a user given their username and password and return an access token"""
    token = await oauth.keycloak_sso.fetch_token(
        username=username,
        password=password,
        grant_type="password",
    )
    return token
```

**Comment by HarryMWinters** [comment]
Fwiw I implemented a solution [here](https://github.com/HarryMWinters/fastapi-oidc).

**Comment by raghavsikaria** [comment]
I tried the Google Authlib method, as mentioned in the description of this issue but for my own OIDC provider. However, FASTAPI gave me an error for the SSL certificate. On debugging I found, when it tries to hit redirect UI for authentication, of my OIDC provider which is "https", it is unable to do so. How can I set SSL verification to False?
